### PR TITLE
TYP/CI: bump mypy&pyright

### PR DESCRIPTION
**Pull request by twoertwein**
Mypy aims to have a monthly release.

### CODE REVIEW

File: pandas/_config/localization.py
Comment (twoertwein) [inline pandas/_config/localization.py:50]: mypy doesn't seem to understand `all(...)` to narrow the type.

File: pandas/_libs/tslibs/timestamps.pyi
Comment (twoertwein) [inline pandas/_libs/tslibs/timestamps.pyi:127]: `pd.Timestamp.replace()` returns a `pd.Timestamp`. Mypy was also expecting it to return a `pd.Timestamp`.

**Comment by twoertwein** [comment]
Marking as draft until mypy 0.950 is available through conda

File: pandas/_libs/tslibs/timestamps.pyi
Comment (simonjayhawkins) [reply pandas/_libs/tslibs/timestamps.pyi:127]: in typeshed, the parent class, `datetime` returns `Self` and alias for an unconstrained typevar.

Comment (twoertwein) [reply pandas/_libs/tslibs/timestamps.pyi:127]: Can we already use Self with 3.8+__future__? If not, we could simply use a TypeVar bound by Timestamp and use it for self and the return value.

Comment (simonjayhawkins) [reply pandas/_libs/tslibs/timestamps.pyi:127]: > Can we already use Self with 3.8+**future**?

not sure. will need to check. in the past we have always needed to create bounded typevars to maintain type for class methods returning same type as self, but maybe this is no longer needed. will review our current usage at some point.

Comment (twoertwein) [reply pandas/_libs/tslibs/timestamps.pyi:127]: Seems to come in python 3.11 https://peps.python.org/pep-0673/ not sure whether it will be/is backported with `__future__`

Comment (twoertwein) [reply pandas/_libs/tslibs/timestamps.pyi]: used a TypeVar

**Comment by twoertwein** [comment]
mypy 0.950 is now available on conda-forge :)

File: pandas/core/arrays/masked.py
Comment (twoertwein) [inline pandas/core/arrays/masked.py:961]: Pyright didn't previously find this. I assume the return value can be any `BaseMaskedArray`.

**Comment by twoertwein** [comment]
doc-test failure seems to be unrelated but mypy&pyright passed :)

@simonjayhawkins

File: pandas/_config/localization.py
Comment (simonjayhawkins) [reply pandas/_config/localization.py:50]: this works, but let's not do this yet! (until we can use it elsewhere)

```
        def all_not_none(arg: Iterable[str | None]) -> TypeGuard[Iterable[str]]:
            return all(x is not None for x in arg)

        if all_not_none(normalized_locale):
```

File: pandas/core/arrays/masked.py
Comment (simonjayhawkins) [reply pandas/core/arrays/masked.py:961]: yep. maybe add type annotations to `_maybe_mask_result` also.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @twoertwein lgtm pending green ci.

File: pandas/core/arrays/masked.py
Comment (simonjayhawkins) [reply pandas/core/arrays/masked.py:961]: can be follow-on. I suspect that `_maybe_mask_result` should probably return the same type, but since we use explicit Subclass constructors this would not be the case if we subclass further.

File: pandas/_config/localization.py
Comment (simonjayhawkins) [reply pandas/_config/localization.py:50]: looking through the code there are a couple of other places this pattern is used that do not use `all_not_none` from `pandas/core/common.py`

maybe change this one here, then in the future we only need the TypeGuard in one place.

```
    try:
        import pandas.core.common as com

        locale.setlocale(lc_var, new_locale)
        normalized_locale = locale.getlocale()
        if com.all_not_none(*normalized_locale):
```

note need to prevent circular import.

Comment (twoertwein) [reply pandas/_config/localization.py:50]: Have there been discussions about using `typing_extensions`? Then we could already use TypeGuard.

Comment (simonjayhawkins) [reply pandas/_config/localization.py:50]: no we don't use typing_extensions but we can import from typing in pandas._typing with a guard so that it resolves to Any in older python versions.

File: pandas/core/arrays/masked.py
Comment (twoertwein) [reply pandas/core/arrays/masked.py:961]: I will leave this out - it would require quite a few ignore comments if we do something like:

```py
    @overload
    def _maybe_mask_result(self, result: ArrayLike, mask: ArrayLike) -> ArrayLike:
        ...

    @overload
    def _maybe_mask_result(
        self, result: tuple[ArrayLike, ArrayLike], mask: ArrayLike
    ) -> tuple[ArrayLike, ArrayLike]:
        ...

    def _maybe_mask_result(
        self, result: ArrayLike | tuple[ArrayLike, ArrayLike], mask: ArrayLike
    ) -> ArrayLike | tuple[ArrayLike, ArrayLike]:
```

For example `FloatingArray` and `BooleanArray` require an ndarray not an ExtensionArray.

**Comment by simonjayhawkins** [comment]
@twoertwein planning to restart the ci here once #47015 is green and merged, so nothing more to do here as far as i'm concerned.

File: pandas/_config/localization.py
Comment (simonjayhawkins) [reply pandas/_config/localization.py:50]: > Have there been discussions about using `typing_extensions`?

As long as `typing_extensions` is not a run-time dependency, I would not be adverse to it being added as a dev dependency.

**Comment by EwoutH** [comment]
Mypy 0.960 was also [released](https://mypy-lang.blogspot.com/2022/05/mypy-0960-released.html) last week

**Pull request merged**
